### PR TITLE
Redesign

### PR DESCRIPTION
**Pull request by bogdan-rosianu**

### CODE REVIEW

File: cmd/rosetta/cli.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+	}
+
+	cliFlagLogsFolder = cli.StringFlag{
+		Name:  "logs-folder",
+		Usage: "Specifies where to save the log files.",
+		Value: "",
+	}
+
+	cliFlagObserveActualShard = cli.UintFlag{
+		Name:  "observe-actual-shard",
```
Comment (bogdan-rosianu) [inline cmd/rosetta/cli.go:52]: observer*

File: cmd/rosetta/cli.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+		Usage: "Specifies the actual shard to observe.",
+		Value: 0,
+	}
+
+	cliFlagObserveProjectedShard = cli.UintFlag{
+		Name:  "observe-projected-shard",
```
Comment (bogdan-rosianu) [inline cmd/rosetta/cli.go:58]: observer*

File: cmd/rosetta/cli.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+		Usage: "Specifies the projected shard to observe.",
+		Value: 0,
+	}
+
+	cliFlagObserver = cli.StringFlag{
+		Name:  "observer",
```
Comment (bogdan-rosianu) [inline cmd/rosetta/cli.go:64]: observer-http-url ?

File: cmd/rosetta/cli.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+	}
+
+	cliFlagChainID = cli.StringFlag{
+		Name:  "chain-id",
+		Usage: "Specifies the Chain ID.",
+		Value: "local-testnet",
+	}
+
+	cliFlagNumShards = cli.UintFlag{
+		Name:  "num-shards",
```
Comment (bogdan-rosianu) [inline cmd/rosetta/cli.go:82]: all the configs (num shards, chain ID, min gas price, min gas limit, gas per data byte) can be fetched `observer/network/config`. 
regarding the genesis block, it's quite a good idea actually to include it as well in that endpoint

File: cmd/rosetta/cli.go (hunk 1 above)
Comment (andreibancioiu) [reply cmd/rosetta/cli.go]: Fixed.

File: cmd/rosetta/cli.go (hunk 3 above)
Comment (andreibancioiu) [reply cmd/rosetta/cli.go]: Fixed.

File: cmd/rosetta/cli.go (hunk 2 above)
Comment (andreibancioiu) [reply cmd/rosetta/cli.go]: Fixed.

File: server/services/operations.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+)
+
+// SupportedOperationTypes is a list of the supported operations.
+var SupportedOperationTypes = []string{
+	opTransfer, opFee, opReward, opScResult, opFeeOfInvalidTx, opGenesisBalanceMovement, opFeeRefund,
+}
+
+var (
+	// opStatusSuccess is the operation status for successful operations.
+	opStatusSuccess = "Success"
```
Comment (bogdan-rosianu) [inline server/services/operations.go:22]: define all vars in a single `var( ) ` group?

File: server/services/operations.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+func indexOperations(operations []*types.Operation) {
+	for index, operation := range operations {
+		operation.OperationIdentifier = indexToOperationIdentifier(index)
+	}
+}
+
+func populateStatusOfOperations(operations []*types.Operation) {
+	for _, operation := range operations {
+		operation.Status = &opStatusSuccess
```
Comment (bogdan-rosianu) [inline server/services/operations.go:40]: can this approach cause issues? you set the point of the opStatusSuccess. is it modified later? if yes, then create a new object ( clone)

File: version/constants.go [hunk 7]
```diff
@@ -0,0 +1,6 @@
+package version
+
+var (
+	RosettaVersion = "v0.1.0"
```
Comment (bogdan-rosianu) [inline version/constants.go:4]: a lot of exported items in the branch that don't have a comment

File: Notes.md [hunk 8]
```diff
@@ -0,0 +1,6 @@
+# Implementation notes
+
+ - We do not support the `related_transactions` property, since it's not feasible to properly filter the related transactions of a given transaction by source / destination shard (with respect to the observed shard).
+ - The endpoint `/block/transaction` is not implemented, since all transactions are returned by the endpoint `/block`.
+ - We chose not to support the optional property `Operation.related_operations`. Although the smart contract results (also known as _unsigned transactions_) form a DAG at the protocol level, operations within a transaction are in a simple sequence.
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply Notes.md:5]: Fixed.

File: cmd/rosetta/cli.go [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+
+	cliFlagGenesisBlock = cli.StringFlag{
+		Name:  "genesis-block",
+		Usage: "Specifies the hash of the genesis block, to be returned by network/status. For mainnet, it must be cd229e4ad2753708e4bab01d7f249affe29441829524c9529e84d51b6d12f2a7.",
+		Value: "0000000000000000000000000000000000000000000000000000000000000000",
+	}
+
+	cliFlagGenesisTimestamp = cli.Int64Flag{
+		Name:  "genesis-timestamp",
+		Usage: "Specifies the timestamp of the genesis block. For mainnet, it must be 1596117600.",
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply cmd/rosetta/cli.go:95]: Good idea! Added.

File: cmd/rosetta/logging.go [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+		logsFolder = currentDirectory
+	}
+
+	err = logger.SetLogLevel(logLevel)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(logsFolder) == 0 {
+	}
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply cmd/rosetta/logging.go:39]: Removed (wonder how it got there).

File: cmd/rosetta/main.go [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+	"net/http"
+	"os"
+	"os/signal"
+	"time"
+
+	"github.com/ElrondNetwork/rosetta/server/factory"
+	"github.com/ElrondNetwork/rosetta/server/provider"
+	"github.com/ElrondNetwork/rosetta/version"
+	"github.com/coinbase/rosetta-sdk-go/server"
+
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply cmd/rosetta/main.go:15]: Resorted.

File: cmd/rosetta/main.go [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+	// Set up signal capturing
+	stop := make(chan os.Signal, 1)
+	signal.Notify(stop, os.Interrupt, os.Kill)
+	<-stop
+
+	shutdownContext, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	_ = httpServer.Shutdown(shutdownContext)
+	_ = httpServer.Close()
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply cmd/rosetta/main.go:101]: I think your explanation is correct, yes. Proxy does this as well, it seems:

https://github.com/ElrondNetwork/elrond-proxy-go/blob/master/cmd/proxy/main.go#L698

File: docker/Rosetta.dockerfile [hunk 13]
```diff
@@ -0,0 +1,17 @@
+FROM golang:1.17.6 as builder
+
+# Clone repositories
+RUN git clone https://github.com/ElrondNetwork/rosetta.git --branch=v0.2.0 --depth=1
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply docker/Rosetta.dockerfile:4]: Good catch! Using main. Let's postpone the tags a little bit.

File: scripts/generate_keys.sh [hunk 14]
```diff
@@ -1,5 +1,5 @@
 KEYS_FOLDER=$1
-OBSERVERS=("0" "1" "2" "metachain")
+OBSERVERS=("0")
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply scripts/generate_keys.sh:2]: Good point. Added parameter.

File: scripts/prepare_host.sh [hunk 15]
```diff
@@ -1,15 +1,14 @@
 BASE_FOLDER=$1
 NETWORKS=("testnet" "devnet" "mainnet")
-SHARDS=("0" "1" "2" "metachain")
+SHARDS=("0")
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply scripts/prepare_host.sh:3]: Good point. Added parameter.

File: server/services/operations.go (hunk 5 above)
Comment (andreibancioiu) [reply server/services/operations.go:22]: Fixed.

File: server/services/operations.go (hunk 6 above)
Comment (andreibancioiu) [reply server/services/operations.go:40]: It's never modified. However, added a TODO to improve this design.

File: version/constants.go (hunk 7 above)
Comment (andreibancioiu) [reply version/constants.go:4]: Added extra few (not all).

**Pull request merged**